builder error: using generics

### Describe the bug

`encode(val.$method()) // Problem here !!!!!!!!!!!!!!!!!!!!!!!!!!!!!1!!!!11!`

### Reproduction Steps

```v
import time

struct Human {
	name string
}

struct Item {
	tag string
}

struct String {
	tag string
}

enum Animal {
	dog
	cat
}

type Entity = Animal | Human | Item | String | string | time.Time

struct SomeGame {
	title  string
	player Entity
	other  []Entity
}

@[sumtype_to: Animal]
fn (ent Entity) json_cast_to_animal() Animal {
	return ent as Animal
}

@[sumtype_to: Human]
fn (ent Entity) json_cast_to_human() Human {
	return ent as Human
}

@[sumtype_to: Item]
fn (ent Entity) json_cast_to_item() Item {
	return ent as Item
}

@[sumtype_to: string]
fn (ent Entity) json_cast_to_string() string {
	return ent as string
}

@[sumtype_to: String]
fn (ent Entity) json_cast_to_other_string() String {
	return ent as String
}

@[sumtype_to: 'time.Time']
fn (ent Entity) json_cast_to_time() time.Time {
	return ent as time.Time
}

fn encode[T](val T) {
	$if T is $sumtype {
		$for method in T.methods {
			if method.attrs.len >= 1 {
				if method.attrs[0].contains('sumtype_to') {
					if val.type_name() == method.attrs[0].all_after('sumtype_to:').trim_space() {
						println(val.$method())
						encode(val.$method()) // Problem here !!!!!!!!!!!!!!!!!!!!!!!!!!!!!1!!!!11!
					}
				}
			}
		}
	} $else $if T is $struct {
	} $else {
		// error('cannot encode value with ${typeof(val).name} type')
	}
}

encode(Entity(Human{'Monke'}))
encode(Entity(Animal.cat))
```

### Expected Behavior

```bash
Human{
    name: 'Monke'
}
cat
```

### Current Behavior

```bash
==================
/tmp/v_1000/copy.8521023635428838776.tmp.c:15353: error: cannot convert 'enum <anonymous>' to 'struct string'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.3 be51143

### Environment details (OS name and version, etc.)

Linux Ubuntu

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -113,6 +113,8 @@ mut:
 	comptime_fields_type             map[string]ast.Type
 	comptime_for_field_value         ast.StructField // value of the field variable
 	comptime_enum_field_value        string // current enum value name
+	comptime_for_method              string // $for method in T.methods {}
+	comptime_for_method_var          string // $for method in T.methods {}; the variable name
 	fn_scope                         &ast.Scope = unsafe { nil }
 	main_fn_decl_node                ast.FnDecl
 	match_exhaustive_cutoff_limit    int = 10

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -115,6 +115,7 @@ mut:
 	comptime_enum_field_value        string // current enum value name
 	comptime_for_method              string // $for method in T.methods {}
 	comptime_for_method_var          string // $for method in T.methods {}; the variable name
+	comptime_values_stack            []CurrentComptimeValues // stores the values from the above on each $for loop, to make nesting them easier
 	fn_scope                         &ast.Scope = unsafe { nil }
 	main_fn_decl_node                ast.FnDecl
 	match_exhaustive_cutoff_limit    int = 10

--- a/vlib/v/checker/comptime.v
+++ b/vlib/v/checker/comptime.v
@@ -282,6 +282,7 @@ fn (mut c Checker) comptime_for(mut node ast.ComptimeFor) {
 					return
 				}
 			}
+			c.push_existing_comptime_values()
 			c.inside_comptime_for_field = true
 			for field in fields {
 				if c.field_data_type == 0 {
@@ -306,8 +307,7 @@ fn (mut c Checker) comptime_for(mut node ast.ComptimeFor) {
 					}
 				}
 			}
-			c.comptime_for_field_var = ''
-			c.inside_comptime_for_field = false
+			c.pop_existing_comptime_values()
 		} else if c.table.generic_type_names(node.typ).len == 0 && sym.kind != .placeholder {
 			c.error('iterating over .fields is supported only for structs and interfaces, and ${sym.name} is neither',
 				node.typ_pos)
@@ -315,6 +315,7 @@ fn (mut c Checker) comptime_for(mut node ast.ComptimeFor) {
 		}
 	} else if node.kind == .values {
 		if sym.kind == .enum_ {
+			c.push_existing_comptime_values()
 			sym_info := sym.info as ast.Enum
 			c.inside_comptime_for_field = true
 			if c.enum_data_type == 0 {
@@ -327,6 +328,7 @@ fn (mut c Checker) comptime_for(mut node ast.ComptimeFor) {
 				c.comptime_fields_type['${node.val_var}.typ'] = node.typ
 				c.stmts(mut node.stmts)
 			}
+			c.pop_existing_comptime_values()
 		} else {
 			c.error('iterating over .values is supported only for enums, and ${sym.name} is not an enum',
 				node.typ_pos)
@@ -337,11 +339,13 @@ fn (mut c Checker) comptime_for(mut node ast.ComptimeFor) {
 		methods_with_attrs := sym.methods.filter(it.attrs.len > 0) // methods with attrs second
 		methods << methods_with_attrs
 
+		c.push_existing_comptime_values()
 		for method in methods {
 			c.comptime_for_method = method.name
 			c.comptime_for_method_var = node.val_var
 			c.stmts(mut node.stmts)
 		}
+		c.pop_existing_comptime_values()
 	} else {
 		c.stmts(mut node.stmts)
 	}
@@ -1053,4 +1057,40 @@ fn (mut c Checker) get_comptime_selector_bool_field(field_name string) bool {
 		'is_enum' { return field_sym.kind == .enum_ }
 		else { return false }
 	}
+}
+
+struct CurrentComptimeValues {
+	inside_comptime_for_field    bool
+	comptime_for_field_var       string
+	comptime_fields_default_type ast.Type
+	comptime_fields_type         map[string]ast.Type
+	comptime_for_field_value     ast.StructField
+	comptime_enum_field_value    string
+	comptime_for_method          string
+	comptime_for_method_var      string
+}
+
+fn (mut c Checker) push_existing_comptime_values() {
+	c.comptime_values_stack << CurrentComptimeValues{
+		inside_comptime_for_field: c.inside_comptime_for_field
+		comptime_for_field_var: c.comptime_for_field_var
+		comptime_fields_default_type: c.comptime_fields_default_type
+		comptime_fields_type: c.comptime_fields_type.clone()
+		comptime_for_field_value: c.comptime_for_field_value
+		comptime_enum_field_value: c.comptime_enum_field_value
+		comptime_for_method: c.comptime_for_method
+		comptime_for_method_var: c.comptime_for_method_var
+	}
+}
+
+fn (mut c Checker) pop_existing_comptime_values() {
+	old := c.comptime_values_stack.pop()
+	c.inside_comptime_for_field = old.inside_comptime_for_field
+	c.comptime_for_field_var = old.comptime_for_field_var
+	c.comptime_fields_default_type = old.comptime_fields_default_type
+	c.comptime_fields_type = old.comptime_fields_type.clone()
+	c.comptime_for_field_value = old.comptime_for_field_value
+	c.comptime_enum_field_value = old.comptime_enum_field_value
+	c.comptime_for_method = old.comptime_for_method
+	c.comptime_for_method_var = old.comptime_for_method_var
 }

--- a/vlib/v/checker/comptime.v
+++ b/vlib/v/checker/comptime.v
@@ -44,6 +44,15 @@ fn (mut c Checker) get_comptime_var_type(node ast.Expr) ast.Type {
 	} else if node is ast.SelectorExpr && c.is_comptime_selector_type(node) {
 		// field_var.typ from $for field
 		return c.comptime_fields_default_type
+	} else if node is ast.ComptimeCall {
+		method_name := c.comptime_for_method
+		left_sym := c.table.sym(c.unwrap_generic(node.left_type))
+		f := left_sym.find_method(method_name) or {
+			c.error('could not find method `${method_name}` on compile-time resolution',
+				node.method_pos)
+			return ast.void_type
+		}
+		return f.return_type
 	}
 	return ast.void_type
 }
@@ -322,6 +331,16 @@ fn (mut c Checker) comptime_for(mut node ast.ComptimeFor) {
 			c.error('iterating over .values is supported only for enums, and ${sym.name} is not an enum',
 				node.typ_pos)
 			return
+		}
+	} else if node.kind == .methods {
+		mut methods := sym.methods.filter(it.attrs.len == 0) // methods without attrs first
+		methods_with_attrs := sym.methods.filter(it.attrs.len > 0) // methods with attrs second
+		methods << methods_with_attrs
+
+		for method in methods {
+			c.comptime_for_method = method.name
+			c.comptime_for_method_var = node.val_var
+			c.stmts(mut node.stmts)
 		}
 	} else {
 		c.stmts(mut node.stmts)

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1565,6 +1565,8 @@ fn (mut c Checker) get_comptime_args(func ast.Fn, node_ ast.CallExpr, concrete_t
 			} else if call_arg.expr is ast.ComptimeSelector
 				&& c.table.is_comptime_var(call_arg.expr) {
 				comptime_args[i] = c.get_comptime_var_type(call_arg.expr)
+			} else if call_arg.expr is ast.ComptimeCall {
+				comptime_args[i] = c.get_comptime_var_type(call_arg.expr)
 			}
 		}
 	}

--- a/vlib/v/checker/tests/comptime_field_selector_not_name_err.out
+++ b/vlib/v/checker/tests/comptime_field_selector_not_name_err.out
@@ -40,3 +40,17 @@ vlib/v/checker/tests/comptime_field_selector_not_name_err.vv:15:12: error: expec
       |               ~~~~
    16 | }
    17 |
+vlib/v/checker/tests/comptime_field_selector_not_name_err.vv:15:10: error: compile time field access can only be used when iterating over `T.fields`
+   13 |         }
+   14 |     }
+   15 |     _ = t.$(f.name)
+      |             ^
+   16 | }
+   17 |
+vlib/v/checker/tests/comptime_field_selector_not_name_err.vv:15:10: error: unknown `$for` variable `f`
+   13 |         }
+   14 |     }
+   15 |     _ = t.$(f.name)
+      |             ^
+   16 | }
+   17 |

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1154,6 +1154,14 @@ fn (mut g Gen) change_comptime_args(func ast.Fn, mut node_ ast.CallExpr, concret
 				if param_typ.nr_muls() > 0 && comptime_args[i].nr_muls() > 0 {
 					comptime_args[i] = comptime_args[i].set_nr_muls(0)
 				}
+			} else if mut call_arg.expr is ast.ComptimeCall {
+				if call_arg.expr.method_name == 'method' {
+					sym := g.table.sym(g.unwrap_generic(call_arg.expr.left_type))
+					// `app.$method()`
+					if m := sym.find_method(g.comptime_for_method) {
+						comptime_args[i] = m.return_type
+					}
+				}
 			}
 		}
 	}

--- a/vlib/v/tests/method_call_resolve_test.v
+++ b/vlib/v/tests/method_call_resolve_test.v
@@ -1,0 +1,46 @@
+struct Human {
+	name string
+}
+
+enum Animal {
+	dog
+	cat
+}
+
+type Entity = Animal | Human
+
+@[sumtype_to: Animal]
+fn (ent Entity) json_cast_to_animal() Animal {
+	return ent as Animal
+}
+
+@[sumtype_to: Human]
+fn (ent Entity) json_cast_to_human() Human {
+	return ent as Human
+}
+
+fn encode[T](val T) {
+	$if T is $sumtype {
+		$for method in T.methods {
+			if method.attrs.len >= 1 {
+				if method.attrs[0].contains('sumtype_to') {
+					if val.type_name() == method.attrs[0].all_after('sumtype_to:').trim_space() {
+						// println(val.$method())
+						encode(val.$method()) // Problem here !!!!!!!!!!!!!!!!!!!!!!!!!!!!!1!!!!11!
+					}
+				}
+			}
+		}
+	} $else $if T is $struct {
+		assert val == Human{
+			name: 'Monke'
+		}
+	} $else $if T is $enum {
+		assert val == Animal.cat
+	}
+}
+
+fn test_main() {
+	encode(Entity(Human{'Monke'}))
+	encode(Entity(Animal.cat))
+}

--- a/vlib/v/tests/method_call_resolve_test.v
+++ b/vlib/v/tests/method_call_resolve_test.v
@@ -25,8 +25,7 @@ fn encode[T](val T) {
 			if method.attrs.len >= 1 {
 				if method.attrs[0].contains('sumtype_to') {
 					if val.type_name() == method.attrs[0].all_after('sumtype_to:').trim_space() {
-						// println(val.$method())
-						encode(val.$method()) // Problem here !!!!!!!!!!!!!!!!!!!!!!!!!!!!!1!!!!11!
+						encode(val.$method())
 					}
 				}
 			}

